evennia/web/webclient/static/webclient/js/webclient_gui.js

```diff
@@ -12,618 +12,258 @@
-            // send a specific oob instruction ["cmdname",[args],{kwargs}]
-            line = line.slice(7);
-            var cmdarr = JSON.parse(line);
-            var cmdname = cmdarr[0];
-            var args = cmdarr[1];
-            var kwargs = cmdarr[2];
-            log(cmdname, args, kwargs);
-            Evennia.msg(cmdname, args, kwargs);
-        } else {
-            input_history.add(line);
-            inputfield.val("");
-            Evennia.msg("text", [line], {});
-        }
+    //
+    // GUI Event Handlers
+    //
+
+    // catch all keyboard input, handle special chars
+    var onKeydown = function (event) {
+        // cycle through each plugin's keydown
+        Object.keys(plugins).forEach( function (plugin) {
+            // does this plugin handle keydown events?
+            if( 'onKeydown' in plugins[plugin] ) {
+                // yes, does this plugin claim this event exclusively?
+                if( plugins[plugin].onKeydown(event) ) {
+                    // 'true' claims this event has been handled so preventDefault and bail on further processing
+                    // order is controlled by plugin javascript tag order in the html
+                    event.preventDefault();
+                    return;
+                }
+            }
+        });
     }
-}
 
-// Opens the options dialog
-function doOpenOptions() {
-    if (!Evennia.isConnected()) {
-        alert("You need to be connected.");
-        return;
-    }
 
-    togglePopup("#optionsdialog");
-}
-
-// Closes the currently open dialog
-function doCloseDialog(event) {
-    var dialog = $(event.target).closest(".dialog");
-    dialog.hide();
-}
-
-// catch all keyboard input, handle special chars
-function onKeydown (event) {
-    var code = event.which;
-    var history_entry = null;
-    var inputfield = $("#inputfield");
-    if (code === 9) {
-      return;
+    // Ask if user really wants to exit session when closing
+    // the tab or reloading the page. Note: the message is not shown
+    // in Firefox, there it's a standard error.
+    var onBeforeUnload = function () {
+        return "You are about to leave the game. Please confirm.";
     }
 
-    //inputfield.focus();
 
-    if (code === 13) { // Enter key sends text
-        doSendText();
-        event.preventDefault();
-    }
-    else if (inputfield[0].selectionStart == inputfield.val().length) {
-        // Only process up/down arrow if cursor is at the end of the line.
-        if (code === 38) { // Arrow up
-            history_entry = input_history.back();
-        }
-        else if (code === 40) { // Arrow down
-            history_entry = input_history.fwd();
-        }
-    }
+    //
+    // Evennia Public Event Handlers
+    //
 
-    if (code === 27) { // Escape key
-        if ($('#helpdialog').is(':visible')) {
-          closePopup("#helpdialog");
-        } else {
-          closePopup("#optionsdialog");
-        }
+    // Handle onLoggedIn from the server
```

Comment:
I think a summary of these handled events and what they expect the plugin to provide should be in the top header of this file.